Issue 228

Taktar opened the pull request:
Fixes #228

What was changed?

Modified Login.js to redirect users to a unified dashboard.

Added a new UnifiedDashboard.js component.

Updated App.js to include the new route.

Why was it changed?
To allow users with multiple roles (system admin, shelter admin, volunteer) to log in once and choose their dashboard.

How was it changed?

Login.js: changed navigate() to go to /unified-dashboard.

App.js: added <Route path="/unified-dashboard" ...>.

Created UnifiedDashboard.js to fetch user permissions and display role-specific navigation options.

strejoarciles commented on client_app/src/App.js:
```diff
@@ -15,51 +15,58 @@ import { ShiftDetails } from "./components/shelter/ShiftDetails";
-  ].includes(location.pathname)) return <NavBarShelterDashboard auth={auth} />;
+  if (
+    [
+      "/shelter-dashboard",
```
Render NavBarShelterDashboard only on shelter-related routes (admin, login, dashboard, shifts, etc.)

strejoarciles commented on client_app/src/App.js:
```diff
@@ -15,51 +15,58 @@ import { ShiftDetails } from "./components/shelter/ShiftDetails";
-  </Router>);
+
+  return (
+    <Router>
```
Set up client-side routing: public routes are defined outside ProtectedRoute,
// while all authenticated views (volunteer/shelter features) are nested under it.

strejoarciles commented on client_app/src/components/authentication/UnifiedDashboard.js:
```diff
@@ -0,0 +1,75 @@
+            Authorization: `Bearer ${JSON.parse(token)}`,
+          },
+        });
+        const data = await response.json();
```
UnifiedDashboard provides a role-aware entry point for users with multiple permissions.
It dynamically fetches the authenticated user's access roles (system, shelter, volunteer)
and conditionally renders dashboard options. Shelter access is further scoped by shelter ID(s),
allowing granular navigation to each associated shelter dashboard.

kate-holdener reviewed (changes requested):
When I go to the home page, I see the original login screen, that offers login based on different roles:
<img width="1402" alt="Screenshot 2025-03-28 at 11 22 21 AM" src="https://github.com/user-attachments/assets/5db9cecc-eae8-42b6-a155-6005d2c69c97" />

Instead, we should  see just a Login screen, that allows the user to submit their credentials. Such a screen is shown below, but remove all role specific references from this screen.
<img width="1515" alt="Screenshot 2025-03-28 at 11 23 52 AM" src="https://github.com/user-attachments/assets/acc2f658-973d-4764-86e8-0dc842152750" />

If the user successfully logs in, they should be redirected to the UnifiedDashboard. The UnifiedDashboard should give the user options to navigate to their role-specific dashboard. 

1. An system admin should see Admin Dashboard, Shelter Dashboard (with a way to select which shelter they want to manage), and Volunteer Dashboard.
2. A shelter admin should see Shelter Dashboard (with a way to select which shelter they want to manage), and Volunteer Dashboard.
3. If a user is neither a shelter admin nor a system admin, they should only see a volunteer dashboard.

Taktar commented:
Updated login screen to remove role-based selection as requested. Now uses a single login component that redirects to /unified-dashboard, which handles role-specific navigation. Also fixed JSX linting errors

kate-holdener commented on client_app/src/components/authentication/UnifiedDashboard.js:
I have developer@slu.edu set up to be a shelter admin (for multiple shelters) and a system admin. I verified that /user_permission returns the correct permissions:
```
{
    "full_access": [
        {
            "resource_type": "shelter",
            "resource_ids": [
                "22222",
                "12345",
                "98776",
                "1234545899"
            ]
        },
        {
            "resource_type": "system",
            "resource_ids": []
        }
    ]
}
```
When I sign in as developer@slu.edu, I only get volunteer dashboard. I should also be seeing admin dashboard and shelter dashboard (with multiple shelter options).

kate-holdener commented on client_app/src/components/authentication/UnifiedDashboard.js:
```diff
@@ -0,0 +1,75 @@
+      try {
+        const response = await fetch(`${SERVER}/user_permission`, {
+          headers: {
+            Authorization: `Bearer ${JSON.parse(token)}`,
```
I'm pretty sure this is incorrect. In other places within the client app, headers are retrieved as: `getAuthHeader();` (this function is defined in `client_app/authentication/getAuthHeader.js`

kate-holdener commented on client_app/src/components/authentication/UnifiedDashboard.js:
```diff
@@ -0,0 +1,75 @@
+          <Card className="mb-3">
+            <Card.Body>
+              <Card.Title>Volunteer</Card.Title>
+              <Button onClick={() => navigate("/volunteer-dashboard")}>Go to Volunteer Dashboard</Button>
```
The color used here does not match the theme used in other places of this client app. Please either use existing css to match the color scheme.  
<img width="1708" alt="Screenshot 2025-03-30 at 7 31 04 AM" src="https://github.com/user-attachments/assets/38f29adc-a54e-4030-ac31-5374878e291d" />

kate-holdener reviewed (changes requested):
@Taktar The login screen is correct now. After I log in, I'm not seeing all the dashboards I expect to seep. Please see and address my detailed comments.

kate-holdener commented on client_app/src/authentication/getAuthHeader.js:
```diff
@@ -2,10 +2,10 @@ import getToken from "./getToken";
+
+  return {
+    "Content-Type": "application/json",
+    Authorization: `Bearer ${JSON.parse(token)}`, //  Cleaned format
```
```suggestion
    Authorization: token`, 
```
I think this like will break our login functionality. The server expects it in a certain format.

kate-holdener commented on client_app/src/components/authentication/Login.js:
```diff
@@ -41,19 +39,20 @@ export default function Login({ setAuth, userRole }) {
     setAuth(true);
-    navigate(userRole === "shelter" ? "/shelter-dashboard" : "/volunteer-dashboard");
-  };  
+    navigate("/unified-dashboard"); // ✅ Use unified dashboard after login
```
```suggestion
    navigate("/unified-dashboard"); // Use unified dashboard after login
```

kate-holdener commented on client_app/src/components/authentication/UnifiedDashboard.js:
```diff
@@ -0,0 +1,77 @@
+    const fetchPermissions = async () => {
+      try {
+        const response = await fetch(`${SERVER}/user_permission`, {
+          headers: getAuthHeader(), // ✅ Use shared helper here
```
```suggestion
          headers: getAuthHeader(), // Use shared helper here
```

kate-holdener reviewed (changes requested):
@Taktar I added a few comments and suggested minor changes.
The big problem is that I get an error when I run your code with 'npm start' (see screenshot below). You should run and test your code before you submit it for review.
<img width="1677" alt="Screenshot 2025-04-02 at 11 22 20 AM" src="https://github.com/user-attachments/assets/2abf7e45-d41e-45f6-b64c-443d8d247592" />

kate-holdener commented:
@Taktar please note that since we merged some pull requests into the main branch, your branch now has merge conflicts with the main. You'll need to resolve those merge conflicts before we can merge your PR.

Taktar commented:
Thank you, I will just take the 0 for this one because I just keep messing up.


[Saint Louis University]

Taslima Aktar she/her
Saint Louis University Class of 2025
College of Arts and Sciences
Computer Science Major


Phone: 314-398-5048

***@***.***
slu.edu<https://www.slu.edu/> | slubillikens.com<https://slubillikens.com/>

________________________________
From: Kate Holdener ***@***.***>
Sent: Wednesday, April 2, 2025 11:25 AM
To: oss-slu/shelter_volunteers ***@***.***>
Cc: Taslima Aktar ***@***.***>; Mention ***@***.***>
Subject: [External] Re: [oss-slu/shelter_volunteers] Issue 228 (PR #239)

@ Taktar please note that since we merged some pull requests into the main branch, your branch now has merge conflicts with the main. You'll need to resolve those merge conflicts before we can merge your PR. — Reply to this email directly,


@Taktar<https://urldefense.com/v3/__https://github.com/Taktar__;!!K543PA!JVY9Poelvw_wuYaKCQ7DWH4a6uh41yAlTuKEbK-kzb7NkikpFI6HZYMA4hqBglWPsa5ds1-czdSSyuU1TXx3oHPJ7B0$> please note that since we merged some pull requests into the main branch, your branch now has merge conflicts with the main. You'll need to resolve those merge conflicts before we can merge your PR.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/239*issuecomment-2773115726__;Iw!!K543PA!JVY9Poelvw_wuYaKCQ7DWH4a6uh41yAlTuKEbK-kzb7NkikpFI6HZYMA4hqBglWPsa5ds1-czdSSyuU1TXx3YDRDt6I$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFOPC4BCSSVOUNC4R7F4O6D2XQFRLAVCNFSM6AAAAABZV5RJ62VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZTGEYTKNZSGY__;!!K543PA!JVY9Poelvw_wuYaKCQ7DWH4a6uh41yAlTuKEbK-kzb7NkikpFI6HZYMA4hqBglWPsa5ds1-czdSSyuU1TXx351O21pI$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

[kate-holdener]kate-holdener left a comment (oss-slu/shelter_volunteers#239)<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/239*issuecomment-2773115726__;Iw!!K543PA!JVY9Poelvw_wuYaKCQ7DWH4a6uh41yAlTuKEbK-kzb7NkikpFI6HZYMA4hqBglWPsa5ds1-czdSSyuU1TXx3YDRDt6I$>

@Taktar<https://urldefense.com/v3/__https://github.com/Taktar__;!!K543PA!JVY9Poelvw_wuYaKCQ7DWH4a6uh41yAlTuKEbK-kzb7NkikpFI6HZYMA4hqBglWPsa5ds1-czdSSyuU1TXx3oHPJ7B0$> please note that since we merged some pull requests into the main branch, your branch now has merge conflicts with the main. You'll need to resolve those merge conflicts before we can merge your PR.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/239*issuecomment-2773115726__;Iw!!K543PA!JVY9Poelvw_wuYaKCQ7DWH4a6uh41yAlTuKEbK-kzb7NkikpFI6HZYMA4hqBglWPsa5ds1-czdSSyuU1TXx3YDRDt6I$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFOPC4BCSSVOUNC4R7F4O6D2XQFRLAVCNFSM6AAAAABZV5RJ62VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZTGEYTKNZSGY__;!!K543PA!JVY9Poelvw_wuYaKCQ7DWH4a6uh41yAlTuKEbK-kzb7NkikpFI6HZYMA4hqBglWPsa5ds1-czdSSyuU1TXx351O21pI$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
